core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -0,0 +1,423 @@
+  }
+
+  case class RedeemScript(redeemScript: ScriptPubKey) extends InputPSBTRecord {
+    override val key: ByteVector = hex"04"
+    override val value: ByteVector = redeemScript.asmBytes
+  }
+
+  case class WitnessScript(witnessScript: ScriptPubKey)
+      extends InputPSBTRecord {
+    override val key: ByteVector = hex"05"
+    override val value: ByteVector = witnessScript.asmBytes
+  }
+
+  case class BIP32DerivationPath(
+      pubKey: ECPublicKey,
+      masterFingerprint: ByteVector,
+      path: BIP32Path)
+      extends InputPSBTRecord {
+    require(pubKey.size == 33, s"pubKey must be 33 bytes, got: ${pubKey.size}")
+
+    override val key: ByteVector = hex"06" ++ pubKey.bytes
+    override val value: ByteVector =
+      path.path.foldLeft(masterFingerprint)(_ ++ _.toUInt32.bytesLE)
+  }
+
+  case class FinalizedScriptSig(scriptSig: ScriptSignature, extra: ByteVector)
+      extends InputPSBTRecord {
+    override val key: ByteVector = hex"07"
+    override val value: ByteVector = scriptSig.bytes ++ extra
+  }
+
+  case class FinalizedScriptWitness(scriptWitness: ScriptWitness)
+      extends InputPSBTRecord {
+    override val key: ByteVector = hex"08"
+    override val value: ByteVector = scriptWitness.bytes
+  }
+
+  case class ProofOfReservesCommitment(porCommitment: ByteVector)
+      extends InputPSBTRecord {
+    override val key: ByteVector = hex"09"
+    override val value: ByteVector = porCommitment
+  }
+
+  case class Unknown(key: ByteVector, value: ByteVector) extends InputPSBTRecord
+}
+
+sealed trait OutputPSBTRecord extends PSBTRecord
+
+object OutputPSBTRecord {
+
+  case class RedeemScript(redeemScript: ScriptPubKey) extends OutputPSBTRecord {
+    override val key: ByteVector = hex"00"
+    override val value: ByteVector = redeemScript.asmBytes
+  }
+
+  case class WitnessScript(witnessScript: ScriptPubKey)
+      extends OutputPSBTRecord {
+
+    override val key: ByteVector = hex"01"
+    override val value: ByteVector = witnessScript.asmBytes
+  }
+
+  case class BIP32DerivationPath(
+      pubKey: ECPublicKey,
+      masterFingerprint: ByteVector,
+      path: BIP32Path)
+      extends OutputPSBTRecord {
+    require(pubKey.size == 33, s"pubKey must be 33 bytes, got: ${pubKey.size}")
+
+    override val key: ByteVector = hex"02" ++ pubKey.bytes
+    override val value: ByteVector =
+      path.path.foldLeft(masterFingerprint)(_ ++ _.toUInt32.bytesLE)
+  }
+
+  case class Unknown(key: ByteVector, value: ByteVector)
+      extends OutputPSBTRecord
+}
+
+sealed trait PSBTMap extends NetworkElement {
+  require(elements.map(_.key).groupBy(identity).values.forall(_.length == 1),
+          "All keys must be unique.")
+
+  def elements: Vector[PSBTRecord]
+
+  def bytes: ByteVector =
+    elements.foldLeft(ByteVector.empty)(_ ++ _.bytes) ++ hex"00"
+}
+
+// Key may be confusing because this is just the first byte of the key
+// Maybe change to Identifier, Byte,  or something else
+sealed trait PSBTKey {
```

Comment:
I like `PSBTKeyId` and then in the scaladoc explain that it is the first byte of a record's key